src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSHost.cs

```diff
@@ -65,11 +65,18 @@ public static SynchronizationContext CurrentOrMainJSSynchronizationContext
             get
             {
 #if FEATURE_WASM_THREADS
-                return JSSynchronizationContext.CurrentJSSynchronizationContext ?? JSSynchronizationContext.MainJSSynchronizationContext!;
+                return (JSProxyContext.ExecutionContext ?? JSProxyContext.MainThreadContext).SynchronizationContext;
 #else
                 return null!;
 #endif
             }
         }
+
+#if FEATURE_WASM_THREADS
+        public static void AssertOperationStack(int expected)
```

Comment:
Can we add new members to public classes in implementation assembly?

Reply:
the whole concept of `PendingOperation` was useful for me to realize all the places which are relevant, but it's too heavy weight. I'm working on another commit which will replace it with new field on `JSMarshalerArgument` and TLS field for context capture.